### ISSUE TITLE
Queueing on unrelated databases after pause/resume

### ISSUE THREAD

**Issue by jcoleman** [pgbouncer/pgbouncer#1145]
We use PAUSE/RESUME (with a settings change and RELOAD in between) to atomically cutover between database hosts.

We've been noticing that we regularly see substantial queuing on database A after we've paused and resumed database B (yes, we're using `PAUSE <database_name>` not a global pause). We've eliminated applications side issues like "had a transaction open on database A while trying to query against database B". The real headscratcher here is that we do _not_ see the queueing if we do not resume database B. Not resuming does, of course, cause other app problems (😁), but it suggests that perhaps there's something about reconnecting the pool for database B that's (substantially?) slowing down activity in other pools. We see this even if there's almost zero activity on database B, so this isn't purely a "lots of client connections are waiting on the paused database" problem. The pool size server side is quite small (~10), but there are thousands of client side connections across many databases.

We're not using PAM (I saw an issue that looked very much like this one...except it was a `usleep` in `pam.c`).

We're using TLS (so I imagine some CPU burn).

We're also currently stuck on 1.17 -- we're trying to upgrade but currently noting random crashes on the newest version, and we're trying to bisect to find the first broken release.

Any ideas on locking or CPU intensive code in the event loop that could cause this behavior?

**Comment by AndrewJackson2020** [comment]
Regarding the crashes you were seeing I assume you are referring to v1.23.0. v1.23.1 was just released which should fix all of that, if not that would be good to know in a a separate issue.

**Comment by cosgroveb** [comment]
We’ve seen the process mysteriously stop at a low frequency per-process (but at our scale a daily occurrence because we run many pgbouncers) on pgbouncer 1.22 first when attempting to upgrade from 1.17 to 1.22.

We haven’t been able to reproduce outside of high throughput deployed environments and while we can’t definitively rule out the environment yet it is notable that on otherwise identical containers (all the same bookworm base image) the lowest tagged release of pgbouncer that exhibits the issue in our envs has proven to be 1.20.1 whereas 1.19.1 is apparently green. We plan to bisect the revisions between the two next which is a slow process for us because of the infrequency and deploying new containers etc

**Comment by AndrewJackson2020** [comment]
Is there an error message to that you are seeing with the process stopping? What you are talking about sounds different from the 1.23.1 patch.

**Comment by cosgroveb** [comment]
That’s the baffling bit. No error message in the logs and that was with the logging level at a 2.

**Comment by cosgroveb** [comment]
I’ll keep you posted about our testing and will open a separate issue if needed. Sounds like no other similar reports though?

In any case this is orthogonal to James’ issue above except insofar as it has prevented us from running a newer pgbouncer

**Comment by cosgroveb** [comment]
Not an OOM but could be a segfault

**Comment by JelteF** [comment]
@jcoleman regarding the queuing. Does the setting change require the existing clients to reconnect (e.g. different hostname or something).

If so it could be that you're hitting into the limitation that PgBouncer only opens a single new connection in parallel at the moment. 

See `launch_new_connection`:

```c
	/*
	 * Allow only a single connection attempt at a time.
	 *
	 * NOTE: If this is ever changed to allow more than a single connection
	 * attempt at once (which would probably be a good thing), some code needs
	 * to change that depends on the fact that there's only ever one connection
	 * attempt at once. At least a little bit below in this function, where
	 * connections are opened for cancel requests.
	 */
	if (!statlist_empty(&pool->new_server_list)) {
		log_debug("launch_new_connection: already progress");
		return;
	}
```

This is definitely a limitation that I would like to see fixed. So if this is indeed the cause a PR is definitely welcome.

**Comment by jcoleman** [comment]
@JelteF I think the problem with that theory is that the pools that are queuing are ones that aren’t being changed (and not being paused and therefore not being reconnected), and the pool that is being paused (and therefore reconnected) can have almost zero activity on it.

**Comment by jcoleman** [comment]
@JelteF Is your mental model of `PAUSE database_name` that it would disconnect all server connections in _only_ the pool `database_name` or do you expect it to _also_ disconnect server connections on e.g. pool `other_database_name`?

**Comment by jcoleman** [comment]
Ooh, I think we've figured this out (HT @micparrish for finding this in the source): `load_config` sets all of the pools to dead, and then `config_postprocess` kills all of those connections -- even if the config hasn't changed for those pools.

**Comment by cosgroveb** [comment]
It appears that every server connection across all pools is marked dirty on every `RELOAD` whether any configs are changing or not. They are then aggressively disconnected. Is this expected behavior? This is somewhat surprising to me as I would expect only servers with new connection strings or new SSL configs to be reconnected.

**Comment by AndrewJackson2020** [comment]
The documentation suggests it should have more nuanced behaviour than what you are describing: "PgBouncer notices when a configuration file reload changes the connection parameters of a database definition. An existing server connection to the old destination will be closed when the server connection is next released (according to the pooling mode), and new server connections will immediately use the updated connection parameters."

https://www.pgbouncer.org/usage.html

**Comment by jcoleman** [comment]
@AndrewJackson2020 Yeah, I think that’s just wrong. The code doesn’t seem to do any comparisons at all on config values. See [load_config](https://github.com/pgbouncer/pgbouncer/blob/aff5a8f49b18d5a07b0fce80a7e471ae972bd781/src/main.c#L431) and then [config_postprocess](https://github.com/pgbouncer/pgbouncer/blob/aff5a8f49b18d5a07b0fce80a7e471ae972bd781/src/janitor.c#L894).

**Comment by cosgroveb** [comment]
@jcoleman based on the poking and additional logging @micparrish and I did today I don't think it's in `load_config` / `config_postprocess` but rather right after `load_config` is called in a call to `sbuf_tls_setup()` here: https://github.com/pgbouncer/pgbouncer/blob/master/src/sbuf.c#L1326-L1333

**Comment by cosgroveb** [comment]
`load_config` is a red-herring. Although it unconditionally marks everything dead on line 431 and you can verify that by logging the contents of the struct immediately after that function call... If you log again after 435 `db_dead` is set back to false already and I am pretty sure the reason for that is that loading the config from the ini file just totally replaces the contents of these structs.

It's definitely `sbuf_tls_setup`, though, whether your config changes TLS or not... Or anything at all. I'm still working on a clean test case to reproduce it but got busy with family stuff at the end of the day.

**Comment by cosgroveb** [comment]
@JelteF this test demonstrates the somewhat surprising behavior we're seeing: https://github.com/pgbouncer/pgbouncer/commit/83142c6138e566bf96589e6556d489e0b4690afa

It kinda makes sense that pgbouncer would do this since it's simplest to just get new connections whenever possible to servers when the config _may or may not have changed_. In our case, however, client poolers are connected to up to 80+ databases and so this is very disruptive. Separately we are addressing the client pooler bottleneck by striping them but we are wondering... 

Is this expected behavior or would it be an improvement if pgbouncer could more granularly diff the actual changes and only refresh connections with actual changes? I can work up a test that demonstrates more closely my expectation here.

**Comment by JelteF** [comment]
@cosgroveb It's definitely on purpose for security reasons, so that you're not stuck with an old TLS config if you changed it. I added this as part of my TLS hot-reloading PR: https://github.com/pgbouncer/pgbouncer/pull/550/files#diff-06d0a51bc262a6a6d6cb682fdada03f4a76ccfb3ed7a09fa969acf1a34d76b82R1043-R1052

But I also completely understand that this can be problematic in its current form. So a PR that changes this logic to only refresh connections when the TLS config actually changed would definitely be something that I would like to merge.

**Comment by cosgroveb** [comment]
Thanks, @JelteF. I'll take a stab at it today.
